File: distribution/api-clients/python/heady_sdk.py
"""
Heady SDK — Python Client
Communicates with heady-manager API from any Python environment

Usage:
    from heady_sdk import HeadyClient
    heady = HeadyClient(api_url="http://manager.dev.local.heady.internal:3300")
    reply = heady.chat("Hello")

Install:
    pip install heady-sdk
    # or copy this file into your project
"""

import json
import urllib.request
import urllib.error
from typing import Any, Dict, List, Optional


class HeadyClient:
    def __init__(self, api_url: str = "http://manager.dev.local.heady.internal:3300", api_key: str = "", timeout: int = 30):
        self.api_url = api_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    def _headers(self) -> Dict[str, str]:
        h = {"Content-Type": "application/json"}
        if self.api_key:
            h["Authorization"] = f"Bearer {self.api_key}"
        return h

    def _request(self, method: str, path: str, body: Optional[Dict] = None) -> Any:
        url = f"{self.api_url}{path}"
        data = json.dumps(body).encode("utf-8") if body else None
        req = urllib.request.Request(url, data=data, headers=self._headers(), method=method)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            raise Exception(f"Heady API {e.code}: {e.reason}")
        except urllib.error.URLError as e:
            raise Exception(f"Cannot reach Heady at {self.api_url}: {e.reason}")

    def _get(self, path: str) -> Any:
        return self._request("GET", path)

    def _post(self, path: str, body: Dict) -> Any:
        return self._request("POST", path, body)

    # ── Health ──────────────────────────────────────────────
    def health(self) -> Dict:
        return self._get("/api/health")

    def pulse(self) -> Dict:
        return self._get("/api/pulse")

    # ── Chat / Buddy ────────────────────────────────────────
    def chat(self, message: str, context: str = "", history: List = None, source: str = "python-sdk") -> str:
        data = self._post("/api/buddy/chat", {
            "message": message,
            "context": context,
            "history": history or [],
            "source": source,
        })
        return data.get("reply") or data.get("message") or ""

    def suggestions(self) -> List[str]:
        data = self._get("/api/buddy/suggestions")
        return data.get("suggestions", [])

    # ── Monte Carlo ─────────────────────────────────────────
    def mc_plan(self, task_type: str) -> Dict:
        return self._post("/api/monte-carlo/plan", {"taskType": task_type})

    def mc_result(self, task_type: str, **kwargs) -> Dict:
        return self._post("/api/monte-carlo/result", {"taskType": task_type, **kwargs})

    def mc_metrics(self, task_type: str = "") -> Dict:
        path = f"/api/monte-carlo/metrics?taskType={task_type}" if task_type else "/api/monte-carlo/metrics"
        return self._get(path)

    # ── Patterns ────────────────────────────────────────────
    def patterns(self, category: str = "", state: str = "") -> Dict:
        params = []
        if category:
            params.append(f"category={category}")
        if state:
            params.append(f"state={state}")
        qs = "&".join(params)
        return self._get(f"/api/patterns?{qs}" if qs else "/api/patterns")

    def observe_pattern(self, category: str, key: str, value: Any) -> Dict:
        return self._post("/api/patterns/observe", {"category": category, "key": key, "value": value})

    # ── Self-Critique ───────────────────────────────────────
    def self_status(self) -> Dict:
        return self._get("/api/self/status")

    def critique(self, **kwargs) -> Dict:
        return self._post("/api/self/critique", kwargs)

    # ── Layers ──────────────────────────────────────────────
    def get_layer(self) -> Dict:
        return self._get("/api/layer")

    def switch_layer(self, layer: str) -> Dict:
        return self._post("/api/layer/switch", {"layer": layer})

    # ── Pricing ─────────────────────────────────────────────
    def pricing_tiers(self) -> Dict:
        return self._get("/api/pricing/tiers")

    def fair_access(self) -> Dict:
        return self._get("/api/pricing/fair-access")

    # ── Webhooks ────────────────────────────────────────────
    def send_webhook(self, event: str, **payload) -> Dict:
        return self._post("/api/webhook/inbound", {"event": event, **payload})


# ── CLI Usage ───────────────────────────────────────────────
if __name__ == "__main__":
    import sys
    client = HeadyClient()

    if len(sys.argv) < 2:
        print("Usage: python heady_sdk.py <command> [args]")
        print("Commands: health, chat <message>, pulse, layers, switch <layer>")
        sys.exit(1)

    cmd = sys.argv[1]

    if cmd == "health":
        print(json.dumps(client.health(), indent=2))
    elif cmd == "pulse":
        print(json.dumps(client.pulse(), indent=2))
    elif cmd == "chat" and len(sys.argv) > 2:
        reply = client.chat(" ".join(sys.argv[2:]))
        print(reply)
    elif cmd == "layers":
        print(json.dumps(client.get_layer(), indent=2))
    elif cmd == "switch" and len(sys.argv) > 2:
        print(json.dumps(client.switch_layer(sys.argv[2]), indent=2))
    else:
        print(f"Unknown command: {cmd}")
        sys.exit(1)

File: distribution/api-clients/js/heady-sdk.js
// HEADY_BRAND:BEGIN
// ╔══════════════════════════════════════════════════════════════════╗
// ║  ██╗  ██╗███████╗ █████╗ ██████╗ ██╗   ██╗                     ║
// ║  ██║  ██║██╔════╝██╔══██╗██╔══██╗╚██╗ ██╔╝                     ║
// ║  ███████║█████╗  ███████║██║  ██║ ╚████╔╝                      ║
// ║  ██╔══██║██╔══╝  ██╔══██║██║  ██║  ╚██╔╝                       ║
// ║  ██║  ██║███████╗██║  ██║██████╔╝   ██║                        ║
// ║  ╚═╝  ╚═╝╚══════╝╚═╝  ╚═╝╚═════╝    ╚═╝                        ║
// ║                                                                  ║
// ║  ∞ SACRED GEOMETRY ∞  Organic Systems · Breathing Interfaces    ║
// ║  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ║
// ║  FILE: distribution/api-clients/js/heady-sdk.js                                                    ║
// ║  LAYER: root                                                  ║
// ╚══════════════════════════════════════════════════════════════════╝
// HEADY_BRAND:END
/**
 * Heady SDK — JavaScript/TypeScript Client
 * Communicates with heady-manager API from any JS environment
 *
 * Usage:
 *   import { HeadyClient } from '@headysystems/heady-sdk';
 *   const heady = new HeadyClient({ apiUrl: 'http://manager.dev.local.headysystems.com:3300' });
 *   const reply = await heady.chat('Hello');
 */

export class HeadyClient {
  constructor({ apiUrl = 'http://manager.dev.local.headysystems.com:3300', apiKey = '', timeout = 30000 } = {}) {
    this.apiUrl = apiUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
    this.timeout = timeout;
  }

  _headers() {
    const h = { 'Content-Type': 'application/json' };
    if (this.apiKey) h['Authorization'] = `Bearer ${this.apiKey}`;
    return h;
  }

  async _fetch(path, opts = {}) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);
    try {
      const res = await fetch(`${this.apiUrl}${path}`, {
        ...opts,
        headers: { ...this._headers(), ...opts.headers },
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`Heady API ${res.status}: ${res.statusText}`);
      return res.json();
    } finally {
      clearTimeout(timer);
    }
  }

  // ── Health ──────────────────────────────────────────────
  async health() {
    return this._fetch('/api/health');
  }

  async pulse() {
    return this._fetch('/api/pulse');
  }

  // ── Chat / Buddy ────────────────────────────────────────
  async chat(message, { context = '', history = [], source = 'sdk' } = {}) {
    const data = await this._fetch('/api/buddy/chat', {
      method: 'POST',
      body: JSON.stringify({ message, context, history, source }),
    });
    return data.reply || data.message || '';
  }

  async suggestions() {
    const data = await this._fetch('/api/buddy/suggestions');
    return data.suggestions || [];
  }

  // ── Monte Carlo ─────────────────────────────────────────
  async mcPlan(taskType) {
    return this._fetch('/api/monte-carlo/plan', {
      method: 'POST',
      body: JSON.stringify({ taskType }),
    });
  }

  async mcResult(taskType, result) {
    return this._fetch('/api/monte-carlo/result', {
      method: 'POST',
      body: JSON.stringify({ taskType, ...result }),
    });
  }

  async mcMetrics(taskType) {
    const path = taskType ? `/api/monte-carlo/metrics?taskType=${taskType}` : '/api/monte-carlo/metrics';
    return this._fetch(path);
  }

  // ── Patterns ────────────────────────────────────────────
  async patterns({ category, state } = {}) {
    const params = new URLSearchParams();
    if (category) params.set('category', category);
    if (state) params.set('state', state);
    return this._fetch(`/api/patterns?${params}`);
  }

  async observePattern(category, key, value) {
    return this._fetch('/api/patterns/observe', {
      method: 'POST',
      body: JSON.stringify({ category, key, value }),
    });
  }

  // ── Self-Critique ───────────────────────────────────────
  async selfStatus() {
    return this._fetch('/api/self/status');
  }

  async critique(data) {
    return this._fetch('/api/self/critique', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  // ── Layers ──────────────────────────────────────────────
  async getLayer() {
    return this._fetch('/api/layer');
  }

  async switchLayer(layer) {
    return this._fetch('/api/layer/switch', {
      method: 'POST',
      body: JSON.stringify({ layer }),
    });
  }

  // ── Browser Sync ────────────────────────────────────────
  async syncTabs(tabs, deviceId) {
    return this._fetch('/api/browser/tabs/sync', {
      method: 'POST',
      body: JSON.stringify({ tabs, deviceId, ts: Date.now() }),
    });
  }

  // ── Pricing / Billing ───────────────────────────────────
  async getPricingTiers() {
    return this._fetch('/api/pricing/tiers');
  }

  async getFairAccess() {
    return this._fetch('/api/pricing/fair-access');
  }

  // ── Webhooks ────────────────────────────────────────────
  async sendWebhook(event, payload) {
    return this._fetch('/api/webhook/inbound', {
      method: 'POST',
      body: JSON.stringify({ event, ...payload }),
    });
  }
}

// CommonJS compat
if (typeof module !== 'undefined') {
  module.exports = { HeadyClient };
}
